commands/ls.go

```diff
@@ -68,3 +82,76 @@ func cmdLs(c *cli.Context) {
 
 	w.Flush()
 }
+
+func parseFilters(filters []string) (FilterOptions, error) {
+	options := FilterOptions{}
+	for _, f := range filters {
+		kv := strings.SplitN(f, "=", 2)
+		key, value := kv[0], kv[1]
+
+		switch key {
+		case "swarm":
+			options.SwarmName = append(options.SwarmName, value)
+		case "driver":
+			options.DriverName = append(options.DriverName, value)
+		case "state":
+			options.State = append(options.State, value)
+		default:
+			return options, fmt.Errorf("Unsupported filter key '%s'", key)
+		}
+	}
+	return options, nil
+}
+
+func filterHosts(hosts []*libmachine.Host, filters FilterOptions) []*libmachine.Host {
+	filteredHosts := []*libmachine.Host{}
```

Comment:
Why not just have one check at the beginning of this function, which does something like:

```go
if len(filters.SwarmName) == 0 && len(filters.DriverName) == 0 && len(filters.State) == 0 {
    return hosts
}
```

It would allow you to get rid of the `len` checks / deeply nested logic below, as well as the final `else` condition.